### ISSUE TITLE
[vmap 函数] 使用 SPU 实现 k 近邻算法

### ISSUE THREAD

**Issue by Candicepan** [secretflow/spu#264]
> 此 ISSUE 为 [隐语开源共建计划（SecretFlow Open Source Contribution Plan，简称 SF OSCP）第二期](https://github.com/orgs/secretflow/discussions/741) [任务 spu#253](https://github.com/secretflow/spu/issues/253)子任务，并完成 assign
> 其他开发者欢迎点击[这里](https://github.com/orgs/secretflow/discussions/741)，选择其他你感兴趣的任务，热烈欢迎社区开发者参与共建～

## 设计思路

- 初始化K近邻模型时，指定近邻的个数（n_neighbors）和权重设置（weights）。
- fit方法用于训练模型，接受特征矩阵X和对应的标签y作为输入，并将它们保存在模型中。
- 计算邻居权重：这里将支持两种权重设置：'uniform'，表示所有邻居样本的权重都相同；'distance'，表示权重与距离的倒数成反比。
- 对单个样本进行预测：对于给定的输入特征，它计算输入样本与所有训练样本的距离，并根据近邻的标签和权重计算预测结果。
- transform方法是预测新的样本集的入口。通过vmap函数将_predict_single_sample应用到每个输入样本上，从而实现对整个输入样本集的预测。

PS：在 API 设计的时候可以参考 sklearn 的 API 预留一下 metric 以及相关的超参数

## 任务介绍

- 任务名称：使用 SPU 实现 k 近邻算法
- 技术方向：SPU/SML
- 任务难度：进阶🌟🌟

## 详细要求
- 安全性（尽量少 reveal）
- 功能性：实现算法的基本功能，包括：
  - 支持 fit，transform
  - 提供配置近邻个数等基本参数
  - 起码一种 weight 设置
- 收敛性：包含 simulator 跑出的实验数据并且证明收敛性/准确性（最好有与明文 sklearn 结果的对比，可详见[参考 PR](https://github.com/secretflow/spu/pull/240)）
- 正确性：请确保提交的代码内容为可以直接运行的
- 代码规范：Python 代码需要使用 black+isort 进行格式化（流水线包含代码规范检查卡点）
- 提交说明：关联该 isuue 并提交代码至 https://github.com/secretflow/spu/tree/main/sml
- 特殊说明：若某个特性有特殊的限制，如需要 FM128，需要更多 fxp 等需要在注释文档中明确说明

## 能力要求
- 熟悉经典的机器学习算法
- 熟悉 JAX 或 NumPy，可以使用 NumPy 实现算法

## 操作说明
- 操作指引：https://www.secretflow.org.cn/docs/spu/latest/en-US/getting_started/tutorials/develop_your_first_mpc_application
- 可参考明文实现：[neighbors/_classification](https://github.com/scikit-learn/scikit-learn/blob/7f9bad99d/sklearn/neighbors/_classification.py#L35)
-  [范例](https://github.com/secretflow/spu/blob/main/sml/linear_model/simple_sgd.py)
-  [参考 PR](https://github.com/secretflow/spu/pull/240)

## 开发须知
以下部分代码请必须增加代码注释，对对应代码模块进行说明，包括：
-  __init__函数的超参数含义
-  fit 的具体算法实现说明
